project/apps/api-gateway/src/users/users.controller.ts

```diff
@@ -1,18 +1,125 @@
-import { Controller, Inject } from '@nestjs/common';
+import {
+  Body,
+  Controller,
+  Delete,
+  Get,
+  Param,
+  Patch,
+  Put,
+  Req,
+  Inject,
+  UsePipes,
+  Query,
+  ForbiddenException,
+  BadRequestException,
+} from '@nestjs/common';
 import { ClientProxy } from '@nestjs/microservices';
+import {
+  UpdateUserDto,
+  userFiltersSchema,
+  UserFiltersDto,
+  updateUserSchema,
+} from '@repo/dtos/users';
+import { ZodValidationPipe } from '@repo/pipes/zod-validation-pipe.pipe';
+import { Request } from 'express';
+import { firstValueFrom } from 'rxjs';
 
 @Controller('users')
+// @UseGuards(AuthGuard) // comment out if we dw auth for now
 export class UsersController {
   constructor(
     @Inject('USER_SERVICE')
-    private readonly userServiceClient: ClientProxy,
+    private readonly usersServiceClient: ClientProxy,
+
+    @Inject('AUTH_SERVICE')
+    private readonly authServiceClient: ClientProxy,
   ) {}
 
-  // @Get()
-  // async getUsers() {}
+  @Get()
+  @UsePipes(new ZodValidationPipe(userFiltersSchema))
+  async getUsers(@Req() req: Request, @Query() filters: UserFiltersDto) {
+    const accessToken = req.cookies['access_token'];
+
+    // Check if user is admin, else forbid access [DELETE once role guard is implemented]
+    const userData = await firstValueFrom(
+      this.authServiceClient.send({ cmd: 'me' }, accessToken),
+    );
+
+    if (userData.role != 'Admin') {
+      throw new ForbiddenException('Access denied.');
+    }
+
+    return this.usersServiceClient.send({ cmd: 'get_users' }, filters);
+  }
+
+  @Get(':id')
+  async getUserById(@Req() req: Request, @Param('id') id: string) {
+    const accessToken = req.cookies['access_token'];
+
+    // Check if admin or user is viewing their own account [DELETE once role guard is implemented]
+    const userData = await firstValueFrom(
+      this.authServiceClient.send({ cmd: 'me' }, accessToken),
+    );
+
+    if (userData.role != 'Admin' && userData.id != id) {
+      throw new ForbiddenException('Access denied.');
+    }
+
+    return this.usersServiceClient.send({ cmd: 'get_user' }, id);
+  }
```

Comment:
for this should we really disallow other (non-admin) users from retrieving another user's data?

one use case i can see here is:
after being matched, i want to be able to click into the other guy's profile to see who i'm doing the question with

Reply:
oh yep, good point